File: src/main/java/org/voegtle/weatherwidget/util/DataFormatter.java
package org.voegtle.weatherwidget.util;

import org.voegtle.weatherwidget.data.WeatherData;
import org.voegtle.weatherwidget.location.WeatherLocation;

import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.Locale;

public class DataFormatter {
  private DecimalFormat numberFormat;

  public DataFormatter() {
    this.numberFormat = (DecimalFormat) NumberFormat.getNumberInstance(Locale.GERMANY);
    this.numberFormat.applyPattern("###.#");
  }


  public String formatTemperatureForActivity(WeatherData data) {
    StringBuilder builder = new StringBuilder();
    builder.append(formatTemperature(data.getTemperature()));
    if (data.getInsideTemperature() != null) {
      builder.append(" / ");
      builder.append(formatTemperature(data.getInsideTemperature()));
    }
    return builder.toString();
  }

  public String formatHumidityForActivity(WeatherData data) {
    StringBuilder builder = new StringBuilder();
    builder.append(formatPercent(data.getHumidity()));
    if (data.getInsideHumidity() != null) {
      builder.append(" / ");
      builder.append(formatPercent(data.getInsideHumidity()));
    }
    return builder.toString();
  }

  public String formatWidgetLine(WeatherLocation location, WeatherData data, boolean detailed) {
    StringBuilder weatherData = new StringBuilder(location.getShortName() + " "
        + formatTemperature(data));
    if (detailed) {
      weatherData.append(" | ");
      weatherData.append(formatPercent(data.getHumidity()));
      if (data.getRainToday() != null) {
        weatherData.append(" | ");
        weatherData.append(formatRain(data.getRainToday()));
      }
    }

    return weatherData.toString();
  }

  public String formatTemperature(WeatherData data) {
    String formattedTemperature;
    Float temperature = data.getTemperature();
    if (temperature != null) {
      formattedTemperature = formatTemperature(temperature);
    } else {
      formattedTemperature = "-";
    }
    return formattedTemperature;
  }

  public String formatTemperature(Float temperature) {
    if (temperature != null) {
      return numberFormat.format(temperature) + "°C";
    } else {
      return "";
    }
  }

  public String formatKwh(Float kwh) {
    if (kwh != null) {
      return numberFormat.format(kwh) + "kWh";
    } else {
      return "";
    }
  }

  public String formatRain(Float rain) {
    if (rain != null) {
      return numberFormat.format(rain) + "l";
    } else {
      return "";
    }
  }

  private String formatPercent(Float val) {
    if (val != null) {
      return numberFormat.format(val) + "%";
    } else {
      return "";
    }
  }

  public String formatWind(Float val) {
    if (val != null) {
      return numberFormat.format(val) + "km/h";
    } else {
      return "";
    }

  }

  public String formatWatt(Float watt) {
    if (watt != null) {
      return numberFormat.format(watt) + "W";
    } else {
      return "";
    }
  }
}

File: src/main/java/org/voegtle/weatherwidget/util/DataFormatter.kt
package org.voegtle.weatherwidget.util

import org.voegtle.weatherwidget.data.WeatherData
import org.voegtle.weatherwidget.location.WeatherLocation
import java.text.DecimalFormat
import java.text.NumberFormat
import java.util.*

class DataFormatter {
  private val numberFormat = NumberFormat.getNumberInstance(Locale.GERMANY) as DecimalFormat
  private val integerFormat = NumberFormat.getNumberInstance(Locale.GERMANY) as DecimalFormat

  init {
    this.numberFormat.applyPattern("#,###.#")
    this.integerFormat.applyPattern("#,###")
  }

  fun formatTemperatureForActivity(data: WeatherData): String {
    val builder = StringBuilder()
    builder.append(formatTemperature(data.temperature))
    if (data.insideTemperature != null) {
      builder.append(" / ").append(formatTemperature(data.insideTemperature))
    }
    return builder.toString()
  }

  fun formatHumidityForActivity(data: WeatherData): String {
    val builder = StringBuilder()
    builder.append(formatPercent(data.humidity))
    if (data.insideHumidity != null) {
      builder.append(" / ").append(formatPercent(data.insideHumidity))
    }
    return builder.toString()
  }

  fun formatWidgetLine(location: WeatherLocation, data: WeatherData, detailed: Boolean): String {
    val weatherData = StringBuilder(location.shortName + " " + formatTemperature(data))
    if (detailed) {
      weatherData.append(" | ")
      weatherData.append(formatPercent(data.humidity))
      data.rainToday?.let {
        weatherData.append(" | ").append(formatRain(it))
      }
    }

    return weatherData.toString()
  }

  fun formatTemperature(data: WeatherData): String {
    return formatTemperature(data.temperature)
  }

  fun formatTemperature(temperature: Float?): String = if (temperature != null) numberFormat.format(temperature) + "°C" else ""

  fun formatKwh(kwh: Float?): String = if (kwh != null) numberFormat.format(kwh) + "kWh" else ""

  fun formatRain(rain: Float?): String = if (rain != null) numberFormat.format(rain) + "l" else ""

  private fun formatPercent(value: Float?): String = if (value != null) numberFormat.format(value) + "%" else ""

  fun formatWind(value: Float?): String = if (value != null) numberFormat.format(value) + "km/h" else ""

  fun formatWatt(watt: Float?): String = if (watt != null) integerFormat.format(watt) + "W" else ""

  fun formatBarometer(barometer: Float?): String = if (barometer != null) integerFormat.format(barometer) + "hPa" else ""

  fun formatSolarradiation(solarradiation: Float?): String = if (solarradiation != null) integerFormat.format(solarradiation) + "W/m²" else ""

  fun formatInteger(value: Float?): String = if (value != null) integerFormat.format(value) else ""

  fun formatDistance(distance: Float?): String = if (distance != null) numberFormat.format(distance) + "km" else ""
}
